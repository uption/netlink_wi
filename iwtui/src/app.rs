use std::time::Duration;

use color_eyre::eyre::Result;
use crossterm::event;
use ratatui::prelude::Backend;
use ratatui::widgets::ListState;
use ratatui::Terminal;

use crate::handler::handle_key_events;
use crate::wifi_info::WifiInfoWorker;

/// Application state.
#[derive(Debug)]
pub struct App {
    pub running: bool,
    pub counter: u8,
    pub wifi_info: WifiInfoWorker,
    pub list_state: ListState,
}

impl App {
    pub fn new() -> Self {
        let wifi_info = WifiInfoWorker::start();

        Self {
            running: true,
            counter: 0,
            wifi_info,
            list_state: ListState::default(),
        }
    }

    /// Run the application main loop.
    pub fn run<B: Backend>(&mut self, terminal: &mut Terminal<B>) -> Result<()> {
        while self.running {
            if self.list_state.selected().is_none() {
                self.list_state.select_first();
            }
            self.draw(terminal)?;
            self.handle_events()?;
        }
        Ok(())
    }

    fn draw<B: Backend>(&mut self, terminal: &mut Terminal<B>) -> Result<()> {
        terminal.draw(|frame| frame.render_widget(self, frame.area()))?;
        Ok(())
    }

    /// Handle terminal events.
    fn handle_events(&mut self) -> Result<()> {
        if !event::poll(Duration::from_millis(100))? {
            return Ok(());
        }
        if let event::Event::Key(key_event) = event::read()? {
            handle_key_events(key_event, self)?;
        }
        Ok(())
    }

    /// Handles the tick event of the terminal.
    pub fn tick(&self) {}

    pub fn quit(&mut self) {
        self.running = false;
    }

    pub fn increment_counter(&mut self) {
        if let Some(res) = self.counter.checked_add(1) {
            self.counter = res;
        }
    }

    pub fn decrement_counter(&mut self) {
        if let Some(res) = self.counter.checked_sub(1) {
            self.counter = res;
        }
    }
}
