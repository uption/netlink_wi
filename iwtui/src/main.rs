use color_eyre::Result;

use crate::app::App;

pub mod app;
pub mod handler;
pub mod ui;
pub mod wifi_info;

fn main() -> Result<()> {
    color_eyre::install()?;
    let mut terminal = ratatui::init();
    let mut app = App::new();

    let app_result = app.run(&mut terminal);
    ratatui::restore();
    app_result
}

// #[cfg(test)]
// mod tests {
//     use super::*;
//     use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
//     use handler::handle_key_events;
//     use insta::assert_snapshot;
//     use ratatui::{backend::TestBackend, Terminal};

//     #[test]
//     fn test_render() {
//         let app = App::default();
//         let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
//         terminal
//             .draw(|frame| frame.render_widget(&app, frame.area()))
//             .unwrap();

//         assert_snapshot!(terminal.backend());
//     }

//     #[test]
//     fn test_handle_key_events() {
//         let mut app = App::default();
//         let key_event = KeyEvent::new(KeyCode::Char('q'), KeyModifiers::NONE);
//         handle_key_events(key_event, &mut app).unwrap();
//         assert!(!app.running);

//         let mut app = App::default();
//         let key_event = KeyEvent::new(KeyCode::Char('c'), KeyModifiers::CONTROL);
//         handle_key_events(key_event, &mut app).unwrap();
//         assert!(!app.running);

//         let key_event = KeyEvent::new(KeyCode::Right, KeyModifiers::NONE);
//         handle_key_events(key_event, &mut app).unwrap();
//         assert_eq!(app.counter, 1);

//         let key_event = KeyEvent::new(KeyCode::Left, KeyModifiers::NONE);
//         handle_key_events(key_event, &mut app).unwrap();
//         assert_eq!(app.counter, 0);
//     }
// }
