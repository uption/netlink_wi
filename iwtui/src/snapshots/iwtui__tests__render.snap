---
source: iwtui/src/main.rs
expression: terminal.backend()
---
"┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Counter App Tutorial ━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓"
"┃                                   Value: 0                                   ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┃                                                                              ┃"
"┗━━━━━━━━━━━━━━━━ Decrement <Left> Increment <Right> Quit <Q> ━━━━━━━━━━━━━━━━━┛"
